### PR TITLE
dirs: rm RunMnt; boot: add vars for early boot env layout; sysconfig: take targetdir arg

### PR DESCRIPTION
**Pull request by anonymouse64**
These variables help eliminate the many instances of filepath.Join we have strewn across the codebase, and will also make it easier for some new, upcoming code in boot to do more cross-checking.

Also, refactor sysconfig to take in a target directory so it doesn't depend on knowledge of the early boot partition layout.

This is stacked on top of https://github.com/snapcore/snapd/pull/8337

### CODE REVIEW

File: dirs/dirs.go [hunk 1]
```diff
@@ -125,6 +125,10 @@ var (
 	FeaturesDir string
 
 	RunMnt string
```
Comment (pedronis) [inline dirs/dirs.go]: so we didn't have the bootloader directories defined, here or /boot.  I think what we likely want to do is move these definitions over the boot, at which point boot using them is a bit less controversial.

sysconfig has also usages, but that one should really take a target rootdir

Comment (pedronis) [reply dirs/dirs.go]: we''ll need a tiny bit of help from dirs, to do this in a not annoying way

Comment (anonymouse64) [reply dirs/dirs.go]: this is all changed now

File: boot/boot.go [hunk 2]
```diff
@@ -40,6 +42,40 @@ const (
...
+	// EarlyWritable is the location of the writable partition during the early
+	// boot (i.e. during the initramfs)
+	EarlyWritable = filepath.Join(EarlyUbuntuData, "system-data")
+)
+
+func init() {
+	// register to change the values of Early* dir values when the global
+	// root dir changes
+	dirs.AddRootDirCallback(func(rootdir string) {
+		EarlyRunMnt = filepath.Join(rootdir, "run/mnt")
```
Comment (pedronis) [inline boot/boot.go]: this duplicates a bit the settings above, you could have a setEarlyDirs helper and then both invoke it with dirs.GlobalDirs and set it as callback

File: boot/boot.go [hunk 3]
```diff
@@ -40,6 +42,40 @@ const (
 	TryingStatus = "trying"
 )
 
+var (
+	// EarlyRunMnt is the rootdir where ubuntu partitions are mounted during the
+	// early boot (i.e. the initramfs)
+	EarlyRunMnt = filepath.Join(dirs.GlobalRootDir, "run/mnt")
```
Comment (pedronis) [inline boot/boot.go]: are they all missing a Dir suffix ?

File: boot/boot.go [hunk 4]
```diff
@@ -40,6 +42,40 @@ const (
 	TryingStatus = "trying"
 )
 
+var (
+	// EarlyRunMnt is the rootdir where ubuntu partitions are mounted during the
```
Comment (pedronis) [inline boot/boot.go]: s/rootdir/directory/  is not a rootdir as we usually mean it

File: boot/boot.go (hunk 3 above)
Comment (pedronis) [reply boot/boot.go]: s/early boot/early user space/   which makes me wonder if boot.Early* alone is confusing or not as way to call these

File: dirs/dirs.go [hunk 5]
```diff
@@ -234,6 +234,12 @@ func SnapModeenvFileUnder(rootdir string) string {
 	return filepath.Join(rootdir, snappyDir, "modeenv")
 }
 
+// AddRootDirCallback registers a callback for whenever the rootdir is changed
```
Comment (pedronis) [inline dirs/dirs.go]: s/rootdir/global root directory (set by SetRootDir)/

File: dirs/dirs.go [hunk 6]
```diff
@@ -234,6 +234,12 @@ func SnapModeenvFileUnder(rootdir string) string {
...
+// to enable updates to variables in other packages that depend on rootdir.
```
Comment (pedronis) [inline dirs/dirs.go]: s/rootdir/its location/

File: sysconfig/sysconfig.go [hunk 7]
```diff
@@ -19,10 +19,16 @@
...
-	// TODO: do we really want this kind of specific dir pointers
-	// or more general ones?
+	// CloudInitSrcDir is where to find the cloud-init data when installing it,
+	// i.e. in early boot install mode it could be something like
+	// filepath.Join(boot.EarlyUbuntuSeed,"data")
 	CloudInitSrcDir string
+
+	// CloudInitTargerRootDir is the root directory where to install cloud-init
+	// data, i.e. in early boot it will be something like boot.EarlyWritable
+	CloudInitTargetDir string
```
Comment (pedronis) [inline sysconfig/sysconfig.go]: let's make this one already not cloud init specific,  let's call it TargetRootDir  and change the doc comment accordingly

File: boot/boot.go (hunk 3 above)
Comment (anonymouse64) [reply boot/boot.go]: yes, I will add Dir suffix to all of them and also change Early -> Initramfs everywhere

File: boot/boot.go [hunk 8]
```diff
@@ -40,6 +42,43 @@ const (
 	TryingStatus = "trying"
 )
 
+var (
```
Comment (pedronis) [inline boot/boot.go]: maybe these should got to their own file initramfs20dirs.go ?  can be in a follow up

File: sysconfig/sysconfig.go [hunk 9]
```diff
@@ -19,10 +19,17 @@
...
-	// or more general ones?
+	// CloudInitSrcDir is where to find the cloud-init data when installing it,
+	// i.e. in early boot install mode it could be something like
+	// filepath.Join(boot.InitramfsUbuntuSeedDir,"data")
 	CloudInitSrcDir string
+
+	// TargetDir is the root directory where to install configure
+	// data, i.e. for cloud-init during the initramfs it will be something like
+	// boot.InitramfsWritableDir
+	TargetDir string
```
Comment (pedronis) [inline sysconfig/sysconfig.go]: I would still call this TargetRootDir and rename other related variable accordingly

File: boot/boot.go (hunk 8 above)
Comment (anonymouse64) [reply boot/boot.go]: sure I will put these in their own file in a followup

File: sysconfig/sysconfig.go (hunk 9 above)
Comment (anonymouse64) [reply sysconfig/sysconfig.go]: do you want this in this PR, or can it be a followup?

Comment (anonymouse64) [reply sysconfig/sysconfig.go]: well actually considering that travis is not triggering anything I will just do this and the other thing now :-/

File: boot/initramfs20dirs.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+var (
+	// InitramfsRunMntDir is the directory where ubuntu partitions are mounted
+	// during the initramfs.
+	InitramfsRunMntDir string
```
Comment (mvo5) [inline boot/initramfs20dirs.go]: :+1: Fwiw, I like the new names

File: sysconfig/cloudinit_test.go [hunk 11]
```diff
@@ -68,11 +71,12 @@ func (s *sysconfigSuite) TestCloudInitInstalls(c *C) {
 
 	err := sysconfig.ConfigureRunSystem(&sysconfig.Options{
 		CloudInitSrcDir: cloudCfgSrcDir,
+		TargetRootDir:   boot.InitramfsWritableDir,
 	})
 	c.Assert(err, IsNil)
 
 	// and did copy the cloud-init files
-	ubuntuDataCloudCfg := filepath.Join(dirs.RunMnt, "ubuntu-data/system-data/etc/cloud/cloud.cfg.d/")
+	ubuntuDataCloudCfg := filepath.Join(boot.InitramfsWritableDir, "etc/cloud/cloud.cfg.d/")
```
Comment (mvo5) [inline sysconfig/cloudinit_test.go]: :heart:

**Pull request merged**